windows/src/engine/keyman32/appint/aiTIP.cpp

```diff
@@ -226,34 +226,52 @@ extern "C" __declspec(dllexport) BOOL WINAPI TIPProcessKey(WPARAM wParam, LPARAM
 	_td->TIPGetContext = ctfunc;
 
   AppContextWithStores *savedContext = NULL;  // I4370   // I4978
-  AppContext *savedContextUsingCore  = NULL; // used for common core
-
+  AppContext *savedContextUsingCore  = NULL;  // used for common core
+  km_kbp_option_item *SavedKBDOptions = NULL; // used for common core
   if (!Updateable) {
     if (isUsingCoreProcessor) {
       savedContextUsingCore = new AppContext();
       _td->app->CopyContext(savedContextUsingCore);
+      SavedKBDOptions = SaveKeyboardOptionsCore(_td->lpActiveKeyboard);
+      // TODO: 5653  ---- Alternate option ---
+      // Instead of allocating memory for each option
+      // just update the changed options. Consider the commented out method UpdateCoreKeyboard
+      // options.
+      /*if (_td->lpActiveKeyboard->lpCoreKeyboardOptions) {
+        UpdateKeyboardOptionsCore(_td->lpActiveKeyboard->lpCoreKeyboardState, _td->lpActiveKeyboard->lpCoreKeyboardOptions);
+      } else {
+        _td->lpActiveKeyboard->lpCoreKeyboardOptions = SaveKeyboardOptionsCore(_td->lpActiveKeyboard);
+      }*/
+      //////// End alternate option
     } else {                                                                                   // I4370
       savedContext = new AppContextWithStores(_td->lpActiveKeyboard->Keyboard->cxStoreArray);  // I4978
       _td->app->SaveContext(savedContext);
     }
   }
+
   BOOL res = ProcessHook();
 
   if (!Updateable) {
     if (isUsingCoreProcessor) {
       if (res) {
         // Reset the context if match found
         _td->app->RestoreContextOnly(savedContextUsingCore);
+        RestoreKeyboardOptionsCore(_td->lpActiveKeyboard->lpCoreKeyboardState, SavedKBDOptions);
+        delete SavedKBDOptions;
+        SavedKBDOptions = NULL;
```

Comment:
Does this leak the pointers internal to this array?

Reply:
Yes it does. I have fixed this by adding a function to help free allocated memory for km_kbp_option_items.
I also added to the unit test infrastructure a check of the heap before and after each unit test to catch memory leaks.  This actually found some other memory leaks but only in other unit tests, not the engine code.